src/sagemaker/workflow/functions.py

```diff
@@ -38,27 +39,39 @@ class Join(Expression):
     Attributes:
         values (List[Union[PrimitiveType, Parameter, Expression]]):
             The primitive type values, parameters, step properties, expressions to join.
-        on_str (str): The string to join the values on (Defaults to "").
+        on (str): The string to join the values on (Defaults to "").
     """
 
     on: str = attr.ib(factory=str)
     values: List = attr.ib(factory=list)
 
+    def __str__(self):
+        """String function for Join"""
+        return json.dumps(self.expr)
```

Comment:
I've converted the json string back to dict at pipeline definition compilation before the overall json dump. Thus, it will work as expected as indicated in the unit test in [def test_pipeline_variable_in_pipeline_definition(sagemaker_session)](https://github.com/aws/sagemaker-python-sdk/pull/2995/files#diff-45b6bf58d20615c818e02533e8149c26838841d272568998cc2b77bbe7e4e547R110)

The reason for this is:
- The overridden __str__ method requires an output of str type
- I've tried to define a type `PipelineVariableString`, make it inherit the str type like the following. However, it brought some weird behaviors that I debugged for whole night and I'm not able to explain. 
    - The weird thing is the __init__ of `PipelineVariableString` would be called twice and the first time a `join_var.expr` was passed in as expected and the second time the `join_var` was passed in which changed the behavior of the __str__ 
    - I feel the inheritance for primitive type may be tricky
 
```
class PipelineVariableString(str):
  def __init__(self, var):
    self.expr = var

class Join(...):
...
  def __str__(self):
    return PipelineVariableString(self.expr)
```